openflow15/action.go

```diff
@@ -0,0 +1,605 @@
+	return nil
+}
+
+// Decode Action types.
+func DecodeAction(data []byte) (Action, error) {
+	t := binary.BigEndian.Uint16(data[:2])
+	var a Action
+	switch t {
+	case ActionType_Output:
+		a = new(ActionOutput)
+	case ActionType_CopyTtlOut:
+		a = new(ActionHeader)
+	case ActionType_CopyTtlIn:
+		a = new(ActionHeader)
+	case ActionType_SetMplsTtl:
+		a = new(ActionMplsTtl)
+	case ActionType_DecMplsTtl:
+		a = new(ActionHeader)
+	case ActionType_PushVlan:
+		a = new(ActionPush)
+	case ActionType_PopVlan:
+		a = new(ActionHeader)
+	case ActionType_PushMpls:
+		a = new(ActionPush)
+	case ActionType_PopMpls:
+		a = new(ActionPopMpls)
+	case ActionType_SetQueue:
+		a = new(ActionSetqueue)
+	case ActionType_Group:
+		a = new(ActionGroup)
+	case ActionType_SetNwTtl:
+		a = new(ActionNwTtl)
+	case ActionType_DecNwTtl:
+		a = new(ActionDecNwTtl)
+	case ActionType_SetField:
+		a = new(ActionSetField)
+	case ActionType_PushPbb:
+		a = new(ActionPush)
+	case ActionType_PopPbb:
+		a = new(ActionHeader)
+	case ActionType_Copy_Field:
+		a = new(ActionCopyField)
+	case ActionType_Meter:
+		a = new(ActionMeter)
+	case ActionType_Experimenter:
+		// For Experimenter message, the length of action should be at least 10 bytes,
+		// including type(2 byte), length(2 byte), vendor(4 byte), and subtype(2 byte)
+		if len(data) < NxActionHeaderLength {
+			return nil, errors.New("the []byte is too short to decode OpenFlow experimenter message")
+		}
+		v := binary.BigEndian.Uint32(data[4:8])
+		if v == NxExperimenterID {
+			a = DecodeNxAction(data)
+		}
+	default:
+		return nil, fmt.Errorf("DecodeAction unknown type: %v", t)
+	}
+	err := a.UnmarshalBinary(data)
+	if err != nil {
+		return a, err
+	}
+	return a, nil
+}
+
+// Action structure for OFPAT_OUTPUT, which sends packets out ’port’.
+// When the ’port’ is the OFPP_CONTROLLER, ’max_len’ indicates the max
+// number of bytes to send. A ’max_len’ of zero means no bytes of the
+// packet should be sent.
+type ActionOutput struct {
+	ActionHeader
+	Port   uint32
+	MaxLen uint16
+	pad    []byte // 6 bytes to make it 64bit aligned
+}
+
+// ofp_controller_max_len
+const (
+	OFPCML_MAX       = 0xffe5 /* maximum max_len value which can be used to request a specific byte length. */
+	OFPCML_NO_BUFFER = 0xffff /* indicates that no buffering should be applied and the whole packet is to be sent to the controller. */
+)
+
+// Returns a new Action Output message which sends packets out
+// port number.
+func NewActionOutput(portNum uint32) *ActionOutput {
+	act := new(ActionOutput)
+	act.Type = ActionType_Output
+	act.Length = act.Len()
+	act.Port = portNum
+	act.MaxLen = OFPCML_NO_BUFFER
+	act.pad = make([]byte, 6)
+	return act
+}
+
+func (a *ActionOutput) Len() (n uint16) {
+	return a.ActionHeader.Len() + 12
+}
+
+func (a *ActionOutput) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(a.Len()))
+	var b []byte
+	n := 0
+
+	b, err = a.ActionHeader.MarshalBinary()
+	copy(data[n:], b)
```

Comment:
Is the error ignored on purpose? If yes, maybe it should use "_" to receive the error?

Reply:
Thanks for your review. The err must not be ignored. I will handle this.